README
Geo-ReadGRIB
============

Geo::ReadGRIB is an Perl module that provides read access to data 
distributed in GRIB files. Specifically, I wrote it to access NOAA Wavewatch 
III marine weather model forecasts. As of version 0.98_1 Geo::ReadGRIB is known to 
support Canadian Meteorological Center's GEM model GRIB files. It may support
many other GRIB variants which use a rectangular lat/long grid but they have not 
been tested. Please notify the maintainers and let them know if it does or doesn't 
support your files.

wgrib.c

Geo::ReadGRIB uses the C program wgrib to retrieve the GRIB file catalog and 
to extract the data. wgrib.c is included in the distribution and will compile 
when you make the module. The resulting executable is named wgrib.exe and 
should install in the same location as ReadGRIB.pm. ReadGRIB.pm will search 
for wgrib.exe at run time and die if it can't find it.

wgrib.c is known to compile and install correctly with Geo::ReadGRIB on 
all common platforms. See the CPAN Testers reports for this module for details.


I've also been able to compile wgrib.c with gcc on Solaris Sparc and i386.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


COPYRIGHT AND LICENCE

Copyright (C) 20006, 2009 by Frank Lyon Cox

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Geo::ReadGRIB.

0.01  Fri Apr 21 05:45:54 2006
	- original version; created by h2xs 1.23 with options
		-X -b 5.6.0 -n Geo::ReadGRIB

0.40  Sat May 27

        - Added a tempfile() method based on File::Tmp to name
          temp filed so had to change Perl version to "use 5.6.1"
        - Each method that uses wgrib.exe now uses tempfile() to
          name temp files and then removes them when done. 
        - Changed tests to use full path to test GRIB file
        - POD doc tweaks	 

0.50  June 12

        - New method extractLaLo() extracts forecast data for a given
          type and time for a range of locations. The locations will be
          all (lat, long) points in the GRIB file inside the rectangular
          area defined by (lat1, long1) and (lat2, long2) where lat1
          >= lat2 and long1 <= long2.  - that is, lat1 is north or lat2
          and long1 is west of long2 (or the same as...)

0.51  July 10 2006

        - Added END block and signal handlers to clean up temp files on
          close or interruptions.
        - Added new method getParam() to return values of selected 
          parameters useful to users.  

0.98  March 6 2009

        - Added Geo::ReadGRIB::Place and Geo::ReadGRIB::PlaceIterator
          classes.
        - extractLaLo() now returns a PlaceIterator object;  
        - Deprecated getCatalog() and getCatalogVerbose(). Getting the 
          offset index for each data type and time is now done during 
          object creation.

0.99  Oct 2009

        - A major bug fix release to support Canadian high res gribs.
        - Add a check for GDS byte 6 = 0. Croak if a not (yet) supported
          grid other than lat/long.
        - Discover scan mode from byte 28. Many changes to support south
          to north scanning. Will also print advisory messages if it 
          detects as yet unsupported modes.
        - Some refactoring and new tests. 
        - extract() now returns a PlaceIterator object like extractLaLo
          does.  

1.0   Oct 2009

        - Some documentation changes and minor adjustments. Verification
          of new GRIB type support justified finally moving to v1.0

1.1   Oct 2009

        - Bug fix release for rt.cpan.org ticket #50820 where new test
          fails for some values only on 64bit uselongdouble Perl 

1.2   Nov 2009

        - Improved extractLaLo() by reading all rows from (la1, lo1) to
          (la2, lo2) into a scalar in one read. This will get more data 
          than needed in many cases, but the temporary increase in memory
          footprint will typically be less than a megabyte. Speed increases
          grater that %35 were measured on large extracts. 

1.21  Nov 2009

        - Add tests and doc to improve "kwalitee" (already 104.35%)
        - Add and improve range testing methods in accord with DRY 
          principle and to remove range test from lalo2offset which 
          is called many times when using extractLaLo()

1.3  Jan 2010

        - Bug fix release for rt.cpan.org ticket #53494 where CMC gribs
          require a different interpretation of PDS octets 19 and 20
        - Added fix to cover changes made for fix

1.4  March 2010

        - extractLaLo() no longer saves extracted data to the object by
          default. The new method backflip() turns this behavior back on.
        - A test was added for backflip()


Makefile.PL
use 5.6.1;
use ExtUtils::MakeMaker;
use Config;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my @ppd;

if ($] >= 5.00503) {
  @ppd = (
    ABSTRACT     => 'Perl extension that gives read access to GRIB weather data files',
    AUTHOR       => 'Frank Cox <frank.l.cox@gmail.com>',
  );
}

eval {system("gcc", "-dumpversion")};
if ($?) {
    print STDERR <<EOF;
    #--------------------------- shucks! ---------------------------------
    #
    # The gcc compiler was not found in you PATH.
    #
    # A C compiler is required by this installer and only gcc is
    # currently supported. 
    #
    # Sorry about that.
    #
    # If you use another compiler and would like this modules maintainers
    # to add support for it, please post a bug by sending email to: 
    #
    #     bug-Geo-ReadGRIB\@rt.cpan.org
    #
    # Please mention the compiler you use and the options required to 
    # compile a simple, single file, C program if you know them. It would 
    # also be helpful if you are able to help with testing. Particularly 
    # if your compiler isn't free and/or the maintainers don't have easy
    # access to your OS.
    #
    #---------------------------------------------------------------------

EOF

    print "//\$Config{cc}: $Config{cc}//\n";
    die "OS unsupported";
}

WriteMakefile(
    NAME         => 'Geo::ReadGRIB',
    VERSION_FROM => 'lib/Geo/ReadGRIB.pm', # finds $VERSION
    PREREQ_PM    => {}, # e.g., Module::Name => 1.1
    LICENSE      => 'perl',
    SKIP         => [qw(static static_lib dynamic_lib)],
    clean        => {FILES => "wgrib.exe"},
    @ppd,
);

sub MY::dynamic {
'
dynamic :: $(INST_LIB)/Geo/wgrib.exe
	@$(NOOP)

$(INST_LIB)/Geo/wgrib.exe: $(C_FILES)
	$(CC) -o $(INST_LIB)/Geo/wgrib.exe wgrib.c
';
}

sub MY::libscan {
    my $path = $_[1];
    return '' if $path =~ /\B\.svn\b/;
    return $path;
}



lib/Geo/ReadGRIB/Place.pm
#
#===============================================================================
#
#         FILE:  Place.pm
#
#  DESCRIPTION:  creates Geo::ReadGRIB::Place objects
#
#        FILES:  ---
#         BUGS:  ---
#        NOTES:  ---
#       AUTHOR:  Frank Lyon Cox (Dr), <frank@pwizardry.com>
#      COMPANY:  Practial Wizardry
#      VERSION:  1.0
#      CREATED:  2/3/2009 10:42:57 PM Pacific Standard Time
#     REVISION:  ---
#===============================================================================

package Geo::ReadGRIB::Place;

use strict;
use warnings;

our $VERSION = 1.0;

#--------------------------------------------------------------------------
#  new( )
#--------------------------------------------------------------------------
sub new {
    my $class = shift;
    my $self = {};
    bless $self, $class;
    return $self;
}

#--------------------------------------------------------------------------
#  thisTime( )
#--------------------------------------------------------------------------
sub thisTime {
    my $self = shift;
    my $arg  = shift;
    $self->{time} = $arg if defined $arg;
    return $self->{time};
}

#--------------------------------------------------------------------------
#  lat( )
#--------------------------------------------------------------------------
sub lat {
    my $self = shift;
    my $arg  = shift;
    $self->{lat} = $arg if defined $arg;
    return $self->{lat};
}

#--------------------------------------------------------------------------
#  long( )
#--------------------------------------------------------------------------
sub long {
    my $self = shift;
    my $arg  = shift;
    $self->{long} = $arg if defined $arg;
    return $self->{long};
}

#--------------------------------------------------------------------------
#  types( )
#
#  returns an array ref of type names
#--------------------------------------------------------------------------
sub types {
    my $self = shift;
    my $arg  = shift;
    push @{$self->{types}}, $arg if defined $arg;
    return $self->{types};
}

#--------------------------------------------------------------------------
#  data( type_name )
#
#  takes a type_name and returns the associated data
#--------------------------------------------------------------------------
sub data {
    my $self = shift;
    my $type = shift;
    my $data = shift;

    $self->{data}->{$type} = $data if defined $data;

    if ( not defined $self->{data}->{$type} ) {
        warn "Place: Not a valid type: $type\n";
    }

    return $self->{data}->{$type};
}


=head1 NAME

Geo::ReadGRIB::Place - Contains the value of a one or more data type at a given 
time and geographic location.

=head1 VERSION

This documentation refers to Geo::ReadGRIB::Place version 1.0 as returned by 
a call to Geo::ReadGRIB::PlaceIterator::Current()


=head1 SYNOPSIS

    use Geo::ReadGRIB;

    $w = new Geo::ReadGRIB "grib-file";
    $w->getFullCatalog;

    print $w->show,"\n";
  
    $plit = $w->extractLaLo(data_type, lat1, long1, lat2, long2, time);
    die $w->getError if $w->getError;

    # $plit is a Geo::ReadGRIB::PlaceIterator

    while ( $place = $plit->current() and $plit->next ) {

        # $place is a Geo::ReadGRIB::Place object

        $time       = $place->thisTime;
        $latitude   = $place->lat;
        $longitude  = $place->long;
        $data_types = $place->types; # an array ref of type names

        $data       = $place->data( data_type );

        # process data...
    }


=head1 DESCRIPTION

Objects of this class are returned by the current() method of a 
PlaceIterator object which itself has been returned by the 
extractLaLo() or extract() methods of a Geo::ReadGRIB object. A place 
object has a unique latitude and longitude for one time and has data 
for one or more data types. 

=head1 METHODS

Objects of this class are read only and all parameters may be
accessed by the following methods.

=over 4

=item $object->new;

=item $object->thisTime;

=item $object->lat;

=item $object->long;

=item $object->types;

=item $object->data(type);

=back

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module. Please report problems through

http://rt.cpan.org

or contact Frank Cox, <frank.l.cox@gmail.com> Patches are welcome.

=head1 AUTHOR

Frank Cox, E<lt>frank.l.cox@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Frank Cox

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.



=cut



1;




lib/Geo/ReadGRIB/PlaceIterator.pm
#
#===============================================================================
#
#         FILE:  PlaceIterator.pm
#
#  DESCRIPTION: A Geo::ReadGRIB::PlaceIterator object contains a collection of
#               Geo::ReadGRIB data for a regular area and has methods to return
#               Geo::ReadGRIB::Place in sorted sequence of locations
#
#        FILES:  ---
#         BUGS:  ---
#        NOTES:  ---
#       AUTHOR:  Frank Lyon Cox (Dr), <frank@pwizardry.com>
#      COMPANY:  Practial Wizardry
#      VERSION:  1.0
#      CREATED:  2/2/2009 12:15:57 PM Pacific Standard Time
#     REVISION:  ---
#===============================================================================

package Geo::ReadGRIB::PlaceIterator;

use strict;
use warnings;
use Geo::ReadGRIB::Place;

our $VERSION = 1.0;

#--------------------------------------------------------------------------
#  new( )
#--------------------------------------------------------------------------
sub new {
    my $class = shift;

    my $self = {};
    bless $self, $class;
    $self->_init;
    return $self;
}


#--------------------------------------------------------------------------
#  _init( )
#--------------------------------------------------------------------------
sub _init {
    my $self = shift;

    $self->{shifted_west_by}   = 0;
    $self->isSorted( undef );
    $self->{place_index}  = 0;
    $self->{place_sorted} = [];
    $self->{x_y} = [];
    $self->{count_of_places}   = 0;
    return;
}

#--------------------------------------------------------------------------
#  first( )
#--------------------------------------------------------------------------
sub first {
    my $self = shift;
    $self->_indexData() if not $self->isSorted;
    $self->_setIndex( 0 );
    return;
}


#--------------------------------------------------------------------------
#  next( )
#--------------------------------------------------------------------------
sub next {
    my $self = shift;
    $self->_indexData() if not $self->isSorted;
    if ( $self->_setIndex == $self->{count_of_places} ) {
        return;
    }
    return $self->_setIndex('+');
}

#--------------------------------------------------------------------------
#  current( )
#
#  returns a Geo::ReadGRIB::Place object
#--------------------------------------------------------------------------
sub current {
    my $self = shift;

    $self->_indexData() if not $self->isSorted;

    my $index = $self->_setIndex;

    my $p = Geo::ReadGRIB::Place->new;

    $p->thisTime( $self->{place_sorted}->[ $index ]->[0] );
    $p->lat( $self->{place_sorted}->[ $index ]->[1] );
    $p->long( $self->{place_sorted}->[ $index ]->[2] );

    if ( not defined $p->thisTime ) {
        return;
    }

    for my $type ( keys %{ $self->{data}->{$p->thisTime}->{$p->lat}->{$p->long} } ) {
        $p->types( $type );
        $p->data( $type, $self->{data}->{$p->thisTime}->{$p->lat}->{$p->long}->{$type} );
    }

    return $p;
}


#--------------------------------------------------------------------------
#  numLong( )
#--------------------------------------------------------------------------
sub numLong {
    my $self = shift;
    $self->_indexData() if not $self->isSorted;
    return $self->{x_y}->[0];
}

#--------------------------------------------------------------------------
#  numLat( )
#--------------------------------------------------------------------------
sub numLat {
    my $self = shift;
    $self->_indexData() if not $self->isSorted;
    return $self->{x_y}->[1];
}

#--------------------------------------------------------------------------
#  _setIndex( )
#--------------------------------------------------------------------------
sub _setIndex {
    my $self  = shift;
    my $index = shift;

    if ( defined $index ) {
        if ( $index eq '+' ) {
            $self->{place_index} = $self->{place_index} + 1;
        }
        else {
            $self->{place_index} = $index;
        }
    }

    if (  $self->{place_index} > $self->{count_of_places} ) {
        return;
    }

    return $self->{place_index};
}

#--------------------------------------------------------------------------
#  isSorted( )
#--------------------------------------------------------------------------
sub isSorted {
    my $self = shift;
    my $bool = shift;
    if ( $bool ) {
        $self->{is_sorted} = 1;
    }
    return $self->{is_sorted};
}

#--------------------------------------------------------------------------
#  _indexData( )
#
#  Inspect the data section and prepare a sorted list of
#  [time, lat, long] 
#
#  -- we also count the number of lats and longs
#
#  Object data is updated
#--------------------------------------------------------------------------
sub _indexData {
    my $self = shift;

    #  Inspect the data section and prepare a sorted list of
    #  [time, lat, long] and an array of times
    #  -- we also count the number of lats and longs

    my ( @tm, $these, $dataType );

    my ( $thisTime ) = keys %{ $self->{data} }; 

    my ( $laC, $loC );
    while ( my ( $time, $laf ) = each %{ $self->{data} } ) {
        $laC = 0;
        while ( my ( $la, $lof ) = each %$laf ) {
            while ( my ( $lo, $dtf ) = each %$lof ) {
                $loC++;
                push @$these, [ $time, $la, $lo ];
                ($dataType) = each %$dtf;
            }
        }
    }

    $self->{x_y} = [  $loC / scalar keys %{ $self->{data}->{$thisTime} },
                      scalar keys %{ $self->{data}->{$thisTime} } ];

    $self->{place_sorted} = $self->_sortThese($these);
    $self->{count_of_places} = $loC;
    $self->isSorted( 1 );
    $self->first;

    return;
}

#--------------------------------------------------------------------------
#  _sortThese( )
#--------------------------------------------------------------------------
sub _sortThese {
    my $self    = shift;
    my $coArray = shift;

    # sort by lat and then by long
    my @psort = sort { 
        $a->[0] <=> $b->[0] || $b->[1] <=> $a->[1] || $a->[2] <=> $b->[2] 
    } @$coArray;

    return \@psort;
}

#--------------------------------------------------------------------------
#  addData( time, la, lo, type, data )
#  
#  Add a unit of data to internal structure
#--------------------------------------------------------------------------
sub addData {
    my $self = shift;
    my $time = shift;
    my $la   = shift;
    my $lo   = shift;
    my $type = shift;
    my $data = shift;

    $self->{data}->{$time}->{$la}->{$lo}->{$type} = $data if defined $data;
    $self->{is_sorted} = undef;

    return $self->{data};
}



1;



#--------------------------------------------------------------------------- 
#  Module Documentation
#---------------------------------------------------------------------------

=head1 NAME

Geo::ReadGRIB::PlaceIterator - Provides methods to iterate through GRIB data
in geographic order and to return Geo::ReadGRIB::Place objects for each location.

=head1 VERSION

This documentation refers to Geo::ReadGRIB::PlaceIterator version 1.0

=head1 SYNOPSIS

    use Geo::ReadGRIB;

    $w = new Geo::ReadGRIB "grib-file";
    $w->getFullCatalog;

    print $w->show,"\n";
  
    $plit = $w->extractLaLo(data_type, lat1, long1, lat2, long2, time);
    die $w->getError if $w->getError;

    # $plit is a Geo::ReadGRIB::PlaceIterator

    for $y ( 0 .. $plit->numLat -1 ) {
        for $x ( 0 .. $plit->numLong -1 ) {
        
            my $place = $plit->current;
            
            # $place is a Geo::ReadGRIB::Place object

            $time       = $place->thisTime;
            $latitude   = $place->lat;
            $longitude  = $place->long;
            $data_types = $place->types; # an array ref of type names
            $data = $place->data(data_type);

            # process data for $x, $y

            $plit->next;
        }
    }


=head1 DESCRIPTION

A PlaceIterator objects let you iterate through places.

Objects of this class are returned by the extractLaLo() or extract() method of a 
Geo::ReadGRIB object. It contains data for one or more data types at one or more 
times for the area extracted. The locations are sorted first by time
then latitude and then by longitude. Methods are provided for sequential access 
to this data.

The first() method sets the iterator index to the most northwest (lat, long) 
pair. The current() method returns a Place object for the current location.
The next() method advances the current location one place in rows from west to 
east and north to south starting in the northwest corner and ending in the 
southeast. After the last item in the southeast corner of the extracted region, 
another call to next() returns B<undef>. 

The convenience methods numLat() and numLong() return the number of latitude and
longitude points respectively. This is handy to, for example, provide the x, y
dimensions for a GD::Image new() method.

=head1 METHODS

=over 4

=item $object->new;

=item $object->first;

=item $object->next;

=item $place = $object->current;

=item $y = $object->numLat;

=item $y = $object->numLong;

=item $y = $object->addData;

=item $y = $object->isSorted;

=back

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module. Please report problems through

http://rt.cpan.org

or contact Frank Cox, <frank.l.cox@gmail.com> Patches are welcome.

=head1 AUTHOR

Frank Cox, E<lt>frank.l.cox@gmail.comE<gt>


=head1 LICENCE AND COPYRIGHT

2009 by Frank Cox

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.


=cut



examples/GDrawGRIB.pl
#! /usr/bin/perl
#--------------------------------------------------------------------------
# GDrawGRIB.pl
#
# Create a animated GIF from Wavewatch III GRIB data.
#--------------------------------------------------------------------------

use strict;
use warnings;

# use lib "../lib";
use Geo::ReadGRIB;
use IO::File;
use Carp;
use GD;

our $VERSION = 1.0;

my $gribFile = shift or help();


#----------------------------------------------
# The following variables are scaled for use 
# with the small Wavewatch III GRIB file 
# akw.HTSGW.grb included with this distribution 
# covering the far north Pacific.
#
# For use with other Wavewatch GRIB files adjust
# latitude longitude pairs, time and data type
# as needed.
# 
# The points ($la1, $lo1) and ($la2, $lo2) define
# the corners of rectangular area. 
#
# Printing the output or the show() and getError()
# methods as shown below will help find these
# values.   
#----------------------------------------------
# ADJUST THESE VARIABLES AS NEEDED
#----------------------------------------------
my $la1  = 65;
my $la2  = 45;
my $lo1  = 159.5;
my $lo2  = 236.5;
my $time = 1142078400;
my $type = 'HTSGW';

my $w = Geo::ReadGRIB->new($gribFile);
$w->getFullCatalog;

print $w->getError,"\n" if $w->getError;
print $w->show;

my $plit = $w->extractLaLo( [$type], $la1, $lo1, $la2, $lo2, $time );

croak $w->getError,"\n" if $w->getError;

#---------------------------------------
# Example: Find the range of data values
#          in this file.
#---------------------------------------

$plit->first;
print "number of lat points: ",$plit->numLat," number of long points ",$plit->numLong,"\n";

my ( $min, $max, %unique ) = ( 1000, 0 );
while ( my $place = $plit->current and $plit->next ) {
    next if $place->data($type) eq 'UNDEF';

    $unique{ $place->data( $type ) }++;

    if ( $place->data( $type ) < $min ) {
        $min = $place->data( $type );
    }

    if ( $place->data( $type ) > $max ) {
        $max = $place->data( $type );
    }
}

print "min: $min, max: $max, unique values: ",scalar keys %unique,"\n";

#---------------------------------------
# Example: create a GIFF image from 
# this data.
#---------------------------------------
my $im = GD::Image->new( $plit->numLong, $plit->numLat );    # (width, height)

my $white = $im->colorAllocate( 255, 255, 255 );
my $black = $im->colorAllocate( 0,   0,   0 );
$im->transparent($white);
$im->interlaced('true');

my $colors = Colors();
my @palette;
foreach my $c (@$colors) {
    $palette[ $c->[3] ] = $im->colorAllocate( $c->[0], $c->[1], $c->[2] );
}

# Set up a mapping between data value ranges and colors using function...
my $theseBands = colorBandMap( $min, $max );  # (min, max)

# Not iterate through the data and draw GIF with colors maped to data values
$plit->first;
for  my $y ( 0 .. $plit->numLat -1 ) {
    for my $x ( 0 .. $plit->numLong -1 ) {
        
        my $place = $plit->current;
        my $td = $place->data( $type );

        if ( $td =~ /UNDEF/ ) {
            # land will be UNDEF, make it black
            $im->setPixel( $x, $y, $black );
        }
#       for optional lat long lines
#       elsif ( $place->long % 30 == 0 or $place->lat % 30 == 0 ) {
#            $im->setPixel( $x, $y, $white );
#       }

        else {
            foreach my $band (@$theseBands) {
                # color this pixle accoding to the  
                # band the data is in
                if ( $td >= $band->[0] and $td <= $band->[1] ) {
                    $im->setPixel( $x, $y, $palette[ $band->[2] ] );
                }
            }
        }

        $plit->next;
    }
}

my $file_name = "out-GDrawGRIB.gif";
my $F = IO::File->new( ">$file_name" ) or croak "can't open $file_name";
binmode $F;
print $F $im->gif;
close $F;


#--------------------------------------------------------------------------
# Colors()
#
# Returns an array of RGB values for a spectrum of colors from blue to red
# and numbers 1-33 to use as names in GD code.
#
# This color palette was developed by Frank Cox in 1999 for use when 
# graphing weather data. 
#--------------------------------------------------------------------------
sub Colors {

   my $rgb =
      # blue
     [[   0,170,255, 1],
      [   0,187,255, 2],
      [   0,204,255, 3],
      [   0,221,255, 4],
      [   0,238,255, 5],
      [   0,255,255, 6],
      # blue green                
      [  17,255,255, 7],
      [  34,255,238, 8],
      [  51,255,221, 9],
      [  68,255,204,10],
      [  85,255,184,11],
      [ 102,255,107,12],
      # green                
      [ 119,255,153,13],
      [ 136,255,136,14],
      [ 153,255,119,15],
      [ 170,255,102,16],
      [ 187,255, 85,17],
      [ 204,255, 68,18],
      [ 221,255, 51,19],
      [ 238,255, 34,20],
      # yellow                
      [ 255,255, 17,21],
      [ 255,255,  0,22],
      [ 255,238,  0,23],
      [ 255,221,  0,24],   # orange
      [ 255,187,  0,25],
      [ 255,153,  0,26],
      # red
      [ 255,119,  0,27],
      [ 255, 85,  0,28],
      [ 255, 51,  0,29],
      [ 255, 17,  0,30],
      [ 230,  0,  0,31],
      [ 196,  0,  0,32],
      [ 162,  0,  0,33]];

   return $rgb;
}

#--------------------------------------------------------------------------
# colorBandMap( min, max)
#
# Takes a minimum and maximum value.
#
# Returns an array mapping data values to color number from the GDraw()
# color set 1-33. Array elements are triplets [band-start, band-end, 
# color-number] like this [[a1,b1,c1],[a2,b2,c2],...] where ax and bx 
# define a range of data values (ax < bx) and cx is the color number. 
#--------------------------------------------------------------------------
sub colorBandMap {

    my ( $mn, $mx ) = @_;

    carp "max ($mx) should be bigger than min ($mn)" if not $mx > $mn;

    my $inc = sprintf "%.2f", ( $mx - $mn ) / 33;

    my $bands = [];

    my ( $lower, $upper ) = ( $mn, $mn + $inc );

    for my $n ( 1 .. 32 ) {
        push @$bands, [ $lower, $upper, $n ];
        $lower = sprintf "%.2f", $upper + .01;
        $upper = sprintf "%.2f", $upper + $inc;
    }

    push @$bands, [ $lower, 1000, 33 ];

    return $bands;
}



#--------------------------------------------------------------------------
# help()
#--------------------------------------------------------------------------
sub help {

   my $help = <<"   END";

   Usage: GDrawGRIB.pl gribfile_name
   Example: GDrawGRIB.pl ../lib/Geo/Sample-GRIB/akw.HTSGW.grb

   END

   print $help;
   exit;
}

__END__



#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#  Application Documentation
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


=head1 NAME

GDrawGRIB.pl - An example program using Geo::ReadGRIB to extract data from a 
Wavewatch III GRIB file and draw a map image with it.

=head1 VERSION

This documentation refers to GDrawGRIB.pl version 1.0


=head1 USAGE

    # The following usage example will work with the small Wavewatch III GRIB 
    # file akw.HTSGW.grb included with this distribution when run from the
    # examples directory of the distribution. The file covers wave height in 
    # the far north Pacific for late March 2006.

    Usage: GDrawGRIB.pl gribfile_name
    Example: GDrawGRIB.pl ../lib/Geo/Sample-GRIB/akw.HTSGW.grb

    # The output is a GIF file: out-GDrawGRIB.gif in the current directory.
    #
    # For use with other Wavewatch GRIB files adjust latitude longitude pairs,
    # time and data type as needed. These will all be found as veriables near
    # the beginning of the program.
    # 
    # The points ($la1, $lo1) and ($la2, $lo2) define the corners of a 
    # rectangular area that must be covered in the file. Likewise, $time must
    # be an included time in epoch seconds and $type will be name of an 
    # included data type.
    #
    # Printing the output or the show() and getError() methods as shown in the 
    # code will help find these. 

=head1 DESCRIPTION

This example program shows how to use Geo::ReadGRIB and its iterator methods to
extract data from a Wavewatch III file and create a GIF map image of the data
at a given time. The default values of variables are set to work with the small 
GRIB file included with the Geo::ReadGRIB distribution. Veriable values can be
adjusted for other Wavewatch GRIB files.

Wavewatch III GRIB files can be found under 

ftp://polar.ncep.noaa.gov/pub/waves/

Look in the latest_run directory for files with the .grb suffix. 

=head1 DIAGNOSTICS

Geo::ReadGRIB may set errors after method calls. These can be seen with the 
getError method. Most common errors will report out of range values for
latitude, longitude, or time. Adjust the $la and $lo and $time variables to 
correct.

=head1 DEPENDENCIES

This program requires the GD module.

=head1 BUGS AND LIMITATIONS

There are no known bugs in this example program.
Please report problems to <Maintainer name(s)>  (<contact address>)
Patches are welcome.

=head1 AUTHOR

Frank Cox  E<lt>frank.l.cox@gmail.comE<gt>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2009 Frank Cox  E<lt>frank.l.cox@gmail.comE<gt> All rights reserved.

This example program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See perldoc perlartistic.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut



